### ISSUE TITLE
Cannot install via pip

### ISSUE THREAD

**Issue by BrenBarn** [visualfabriq/bquery#57]
Trying to install via pip (at least on Windows) gives a long error message ending with:

    ValueError: path '/Users/carstvaartjes/PyCharmProjects/bquery/bquery/ctable_ext.pyx' cannot be absolute

This is likely due to the same sort of misconfigured `setup.py` that led to the same issue in bcolz: Blosc/bcolz#266.

**Comment by CarstVaartjes** [comment]
Thanks for the input. Made a new version (0.1.0.1), let me know if that works for you

**Comment by BrenBarn** [comment]
Thanks for the quick reply!  I now get a different error though:

>    build\lib.win-amd64-2.7\bquery\ctable_ext.pyd : fatal error LNK1120: 1 unresolved externals
 >   error: Command "C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\BIN\amd64\link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:C:\FakeProgs\Python27\libs /LIBPATH:C:\FakeProgs\Python27\PCbuild\amd64 /EXPORT:initctable_ext build\temp.win-amd64-2.7\Release\bquery\ctable_ext.obj /OUT:build\lib.win-amd64-2.7\bquery\ctable_ext.pyd /IMPLIB:build\temp.win-amd64-2.7\Release\bquery\ctable_ext.lib /MANIFESTFILE:build\temp.win-amd64-2.7\Release\bquery\ctable_ext.pyd.manifest /MANIFEST" failed with exit status 1120

**Comment by CarstVaartjes** [comment]
Hmmm and it does work when building it from the source? (i do not have that much experience with windows builds myself unfortunately...)

**Comment by BrenBarn** [comment]
Ah, actually it does not work when building from source.  I get the same error.

I realized the previous couple lines of the output may be relevant, here it is:

>    ctable_ext.obj : warning LNK4197: export 'initctable_ext' specified multiple times; using first specification
       Creating library build\temp.win-amd64-2.7\Release\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable_ext.lib and object build\temp.win-amd64-2.7\Release\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable_ext.exp
    ctable_ext.obj : error LNK2019: unresolved external symbol isnan referenced in function __pyx_fuse_0__pyx_f_6bquery_10ctable_ext_count_unique
    build\lib.win-amd64-2.7\bquery\ctable_ext.pyd : fatal error LNK1120: 1 unresolved externals
    error: Command "C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\BIN\amd64\link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:C:\FakeProgs\Python27\libs /LIBPATH:C:\FakeProgs\Python27\PCbuild\amd64 /EXPORT:initctable_ext build\temp.win-amd64-2.7\Release\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable_ext.obj /OUT:build\lib.win-amd64-2.7\bquery\ctable_ext.pyd /IMPLIB:build\temp.win-amd64-2.7\Release\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable_ext.lib /MANIFESTFILE:build\temp.win-amd64-2.7\Release\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable_ext.pyd.manifest /MANIFEST" failed with exit status 1120

The "unresolved external symbol isnan" error is similar to one that was apparently handled in sklearn (scikit-learn/scikit-learn#3566).  Perhaps that discussion/fix will be useful.

**Comment by BrenBarn** [comment]
Also this rather old (resolved) numpy bug: numpy/numpy#2099.

**Comment by BrenBarn** [comment]
Doing some more experimenting, I tried changing [the `nan` import](https://github.com/visualfabriq/bquery/blob/5ea5ee1e727ed3fd937a7eaaad3384cdb3a289ce/bquery/ctable_ext.pyx#L17) in ctable_ext.pyx from `from libc.math import isnan` to `from numpy cimport isnan`.  This seems to be what pandas does (i.e., it uses numpy isnan, not one from libc).  This allows the module to compile, but it breaks one test, `test_groupby_08`.  This is the nose output:

```
...........E.............
======================================================================
ERROR: test_groupby_08: Groupby's type 'count_distinct'
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\FakeProgs\Python27\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\tests\test_ctable.py", line 519, in test_groupby_08
    result_bcolz = fact_bcolz.groupby(groupby_cols, agg_list)
  File "C:\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable.py", line 227, in groupby
    bool_arr=bool_arr)
  File "C:\Users\BrenBarn\Documents\Python\extensions\bquery\bquery\ctable.py", line 162, in aggregate_groups
    raise e
IndexError: arrays used as indices must be integer (or boolean)
-------------------- >> begin captured stdout << ---------------------
TestCtable.setup
data
[('a', 0.0, 0L, 0, nan, -366L, 70) ('a', 0.1, 0L, 0, 0.0, 264L, -49)
 ('b', 1.0, 1L, 1, nan, -5L, -10) ..., ('c', 3.0, 3L, 3, 1.0, 264L, -12)
 ('c', 3.0, 3L, 3, 3.0, 140L, -81) ('c', 3.0, 3L, 3, 1.0, 465L, -21)]
--> Bcolz

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 25 tests in 5.552s

FAILED (errors=1)
```

I don't know enough about bquery/bcolz internals to know exactly what the deal is here, but it seems like maybe it's almost working.

I found another related bug in [astropy](https://github.com/astropy/astropy/pull/186) which references the same numpy bug I mentioned above.  It seems like the basic problem is that trying to get `isnan` from libc is not cross-platform-safe, because `isnan` is not defined in a consistent place across different platform math libraries.  I gather from these other bug reports that the safe thing is to import `isnan` from numpy itself.

**Comment by CarstVaartjes** [comment]
ok, will check why the test fails!

**Comment by CarstVaartjes** [comment]
And thanks for the research and work!

**Comment by CarstVaartjes** [comment]
I changed the isnan to np.isnan and for me all the tests work now, but I only have OSX & Linux environments available to me. It seems that the count distinct doesn't handle nans correctly on Windows then, which is slightly weird to be honest (I would expect the sum with NAs also to fail then, not just the count distinct)

**Comment by CarstVaartjes** [comment]
Ok, after doing some more checking: the count_distinct seems to be a bit dodgy in terms of implementation (using an escape clausule into a different path than all the other aggregation methods). So a few things:

- I will make a new pip installable that will work for Windows right now with the numpy isnan. 
- Windows users should not use count_distinct yet; all other aggregations are correct
- I will remake the count distinct into something consistent with the rest of the code some time soon

I hope that's okay for you

**Comment by CarstVaartjes** [comment]
Okay, a new version is available: 0.10.0.2

**Comment by BrenBarn** [comment]
Okay, I was wondering if that failure was really related.  Thanks for putting in the fix!  It now installs successfully with pip, so I'll close this.
